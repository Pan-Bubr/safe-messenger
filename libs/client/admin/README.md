# client-admin

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test client-admin` to execute the unit tests.
