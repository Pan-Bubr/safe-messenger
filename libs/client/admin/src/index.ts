export * from './lib/client-admin.module';
