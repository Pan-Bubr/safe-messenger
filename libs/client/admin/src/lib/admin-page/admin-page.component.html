<div class="pt-3 pb-3">
  <div class="text-right">
    <button (click)="logout()" class="btn btn-secondary">
      Log out
    </button>
  </div>
</div>

<div>
  <form [formGroup]="addForm" (ngSubmit)="addUser(addForm.value)">
    <div
      *ngIf="
        addForm.controls?.email?.errors?.email &&
        addForm.controls?.displayName?.touched
      "
      class="alert alert-warning"
      role="alert"
    >
      Email invalid
    </div>

    <div class="input-group mb-3">
      <div class="input-group-prepend">
        <span class="input-group-text">New User:</span>
      </div>
      <input
        name="email"
        formControlName="email"
        type="text"
        class="form-control"
        placeholder="Enter email..."
        required
      />
      <input
        name="displayName"
        formControlName="displayName"
        type="text"
        class="form-control"
        placeholder="Enter display name..."
        required
      />

      <div class="input-group-append">
        <button
          class="btn btn-primary"
          type="submit"
          [disabled]="!addForm.valid"
        >
          Add User
        </button>
      </div>
    </div>
  </form>
</div>

<div class="table-responsive-sm">
  <table class="table">
    <thead>
      <tr>
        <th class="col">E-mail</th>
        <th class="col">Delete</th>
      </tr>
    </thead>
    <tbody>
      <tr *ngFor="let email of emails">
        <td class="align-middle">{{ email }}</td>
        <td>
          <button class="btn btn-warning" (click)="delete(email)">
            Delete&nbsp;user
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
