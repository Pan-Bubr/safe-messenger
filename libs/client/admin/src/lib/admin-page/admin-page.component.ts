import { Component, OnDestroy, OnInit } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { Router } from '@angular/router';

import { AdminAuthService } from '../services/admin-auth.service';
import { AdminService } from '../services/admin.service';

@Component({
  selector: 'safe-messenger-admin-page',
  templateUrl: './admin-page.component.html',
  styleUrls: ['./admin-page.component.scss']
})
export class AdminPageComponent implements OnInit, OnDestroy {
  constructor(
    private readonly adminService: AdminService,
    private readonly adminAuthService: AdminAuthService,
    private readonly router: Router
  ) {}

  public newUser: { email: string; displayName: string } = {
    email: '',
    displayName: ''
  };

  public addForm: FormGroup = new FormGroup({
    email: new FormControl(this.newUser.email, [
      Validators.required,
      Validators.email
    ]),
    displayName: new FormControl(this.newUser.displayName, [
      Validators.required
    ])
  });

  public emails: string[] = [];
  public ngOnInit(): void {
    this.adminService.userList().subscribe(emails => {
      this.emails = emails;
    });
  }

  public refresh(): void {
    this.adminService.userList().subscribe(emails => {
      this.emails = emails;
    });
  }

  public addUser({
    email,
    displayName
  }: {
    email: string;
    displayName: string;
  }): void {
    this.adminService.addUser(email, displayName).subscribe(() => {
      this.refresh();
    });

    this.addForm.reset();
  }

  public delete(email: string): void {
    this.adminService.remove(email).subscribe(() => {
      this.refresh();
    });
  }

  public async logout(): Promise<void> {
    this.adminAuthService.logout();
    await this.router.navigateByUrl('/');
  }

  public ngOnDestroy(): void {
    this.adminAuthService.logout();
  }
}
