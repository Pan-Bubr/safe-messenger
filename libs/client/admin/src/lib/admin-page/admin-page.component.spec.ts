import { HttpClientTestingModule } from '@angular/common/http/testing';
import { Router } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';
import { ClientAdminModule } from '@safe-messenger/client/admin';

import { of } from 'rxjs';
import { Shallow } from 'shallow-render';

import { AdminAuthService } from '../services/admin-auth.service';
import { AdminService } from '../services/admin.service';
import { AdminPageComponent } from './admin-page.component';

describe('LoginContainerComponent', () => {
  let shallow: Shallow<AdminPageComponent>;

  beforeEach(() => {
    shallow = new Shallow(AdminPageComponent, ClientAdminModule)
      .import(HttpClientTestingModule)
      .provideMock(
        {
          provide: AdminService,
          useValue: {
            userList: jest.fn(() => of(['email'])),
            addUser: jest.fn(() => of('')),
            remove: jest.fn(() => of(''))
          }
        },
        {
          provide: AdminAuthService,
          useValue: {
            logout: jest.fn()
          }
        },
        {
          provide: Router,
          useValue: {
            navigateByUrl: jest.fn()
          }
        }
      );
  });

  it('should match snapshot', async () => {
    const { fixture } = await shallow.render();

    expect(fixture).toMatchSnapshot();
  });

  it('refresh() should request userList', async () => {
    const { instance, get } = await shallow.render();

    instance.refresh();

    expect(get(AdminService).userList).toBeCalled();
  });

  it('addUser() should request userList', async () => {
    const { instance, get } = await shallow.render();

    instance.addUser({
      email: 'mockEmail',
      displayName: 'mockDisplayName'
    });

    expect(get(AdminService).addUser).toBeCalled();
  });

  it('delete() should request userList', async () => {
    const { instance, get } = await shallow.render();

    instance.delete('mockEmail');

    expect(get(AdminService).remove).toBeCalled();
  });

  it('logout() should request userList', async () => {
    const { instance, get } = await shallow.render();

    await instance.logout();

    expect(get(AdminAuthService).logout).toBeCalled();
  });
});
