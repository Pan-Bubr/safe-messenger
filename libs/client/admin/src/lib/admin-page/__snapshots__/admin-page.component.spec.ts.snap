// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginContainerComponent should match snapshot 1`] = `
<ng-component>
  <safe-messenger-admin-page>
    <div
      class="pt-3 pb-3"
    >
      <div
        class="text-right"
      >
        <button
          class="btn btn-secondary"
        >
           Log out 
        </button>
      </div>
    </div>
    <div>
      <form
        class="ng-untouched ng-pristine ng-invalid"
        novalidate=""
      >
        
        <div
          class="input-group mb-3"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              New User:
            </span>
          </div>
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            formcontrolname="email"
            name="email"
            placeholder="Enter email..."
            required=""
            type="text"
          />
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            formcontrolname="displayName"
            name="displayName"
            placeholder="Enter display name..."
            required=""
            type="text"
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-primary"
              disabled=""
              type="submit"
            >
               Add User 
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      class="table-responsive-sm"
    >
      <table
        class="table"
      >
        <thead>
          <tr>
            <th
              class="col"
            >
              E-mail
            </th>
            <th
              class="col"
            >
              Delete
            </th>
          </tr>
        </thead>
        <tbody>
          
          <tr>
            <td
              class="align-middle"
            >
              email
            </td>
            <td>
              <button
                class="btn btn-warning"
              >
                 DeleteÂ user 
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </safe-messenger-admin-page>
</ng-component>
`;
